### PR TITLE
Partial Assembly for H(curl) and H(div) [pahcurlhdiv-dev]

### PR DESCRIPTION
**Pull request by wang84**
HPC partial assembly for H(curl) and H(div) forms.

TODO:
- [ ] Update CHANGELOG

### CODE REVIEW

**Comment by tzanio** [comment]
@wang84: there are some testing errors, can you take a look?

https://travis-ci.org/mfem/mfem/builds/283887954

File: fem/bilinearform.cpp [hunk 1]
```diff
@@ -27,9 +27,19 @@ void BilinearForm::AllocMat()
       return;
    }
 
-   const Table &elem_dof = fes->GetElementToDofTable();
+   Table elem_dof(fes->GetElementToDofTable());
```
Comment (jdahm) [inline fem/bilinearform.cpp]: I see that the vast majority of the changes are isolated in the templated code, which is great. Out of curiosity, why does BilinearForm have to change?

Comment (wang84) [reply fem/bilinearform.cpp]: This is fixing the seg fault when using UsePrecomputedSparsity() for ND and RT element. It is not for partial assembly.

**Comment by barker29** [comment]
This looks very nice! Have you done any timing comparisons?

File: miniapps/performance/ex3.cpp [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+      {
+         a_hpc->RecoverFEMSolution(X, *b, x);
+      }
+      else
+      {
+         a_h->RecoverFEMSolution(X, *b, x);
+      }
+   }
+
+   // 12. Save the refined mesh and the solution. This output can be viewed
```
Comment (barker29) [inline miniapps/performance/ex3.cpp]: Is there a reason the GLVis save and visualization are commented out?

**Comment by barker29** [comment]
I have a very basic proof of concept LOR preconditioner working in ex3.cpp, but I think it would be much more interesting in Hypre (ex3p.cpp). @wang84 , are you working on a partially assembled ex3p or should I go ahead and take a look myself?

**Comment by wang84** [comment]
Yes. I am working on it right now. I can discuss with you when I have a draft.

Best,

Lu WANG
Center for Applied Scientific Computing
Lawrence Livermore National Laboratory
Livermore, CA 94550
Mail Code: L-561
Phone 925-422-0876

**Comment by tzanio** [comment]
Very nice work @wang84 and @barker29 -- it will be *really* cool to have matrix-free ex3p with LOR AMS!

File: fem/tbilininteg.hpp [hunk 3]
```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
    }
 };
 
+// Mass kernel for Nedelec Finite Element
+
+// complex_t - type for the assembled data
+template <int SDim, int Dim, typename complex_t>
+struct THcurlMassKernel;
```
Comment (YohannDudouit) [inline fem/tbilininteg.hpp]: Is there a reason why you need a new mass Kernel? Nedelec Finite Element are just a different basis, the standard mass Kernel should work no?

File: fem/tbilininteg.hpp [hunk 4]
```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
...
+   // Jt        [M x Dim x SDim x NE] - Jacobian transposed, data member in F
+   // Q                               - CoefficientEval<>::Type
+   // q                               - CoefficientEval<>::Type::result_t
+   // val_qpts [M x SDim x NC x NE]  - in/out data member in R
+   //
+   // val_qpts = (w/det(J)) adj(J) adj(J)^t val_qpts
+   template <typename T_result_t, typename Q_t, typename q_t,
+             typename S_data_t>
+   static inline MFEM_ALWAYS_INLINE
+   void Action(const int k, const T_result_t &F,
```
Comment (YohannDudouit) [inline fem/tbilininteg.hpp]: I don't understand why the mass Kernel would have this form: val_qpts = (w/det(J)) adj(J) adj(J)^t val_qpts. Isn't there a confusion with the Laplacian or some other operator involving derivatives?

File: fem/tbilininteg.hpp [hunk 5]
```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
...
+   // stored as symmetric (when asm_type == p_asm_data, i.e. A.layout.rank == 2)
+   // or non-symmetric (when asm_type == f_asm_data, i.e. A.layout.rank == 3)
+   // matrices.
+   // Jt   [M x Dim x SDim x NE] - Jacobian transposed, data member in F
+   // Q                          - CoefficientEval<>::Type
+   // q                          - CoefficientEval<>::Type::result_t
+   // A    [M x Dim*(Dim+1)/2]   - partially assembled Dim x Dim symm. matrices
+   // A    [M x Dim x Dim]       - partially assembled Dim x Dim matrices
+   //
+   // A = (w/det(J)) adj(J) adj(J)^t
```
Comment (YohannDudouit) [inline fem/tbilininteg.hpp]: same remark as for Action

File: fem/tbilininteg.hpp [hunk 6]
```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
...
+   };
+
+   // Method used for un-assembled (matrix free) action.
+   // Jt        [M x Dim x SDim x NE] - Jacobian transposed, data member in F
+   // Q                               - CoefficientEval<>::Type
+   // q                               - CoefficientEval<>::Type::result_t
+   // val_qpts [M x SDim x NC x NE]  - in/out data member in R
+   //
+   // curl_qpts = (w/det(J))J val_qpts
+   template <typename T_result_t, typename Q_t, typename q_t,
```
Comment (YohannDudouit) [inline fem/tbilininteg.hpp]: Same remark as for Hcurl, is there a reason why the coefficients are curl_qpts = (w/det(J))J val_qpts, and not simply w*det(J)?

File: fem/tbilininteg.hpp [hunk 7]
```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
...
+      {
+         typedef typename T_result_t::Jt_type::data_type real_t;
+         const real_t J11 = F.Jt(i,0,0,k);
+         const real_t J12 = F.Jt(i,1,0,k);
+         const real_t J21 = F.Jt(i,0,1,k);
+         const real_t J22 = F.Jt(i,1,1,k);
+         const complex_t w_det_J = Q.get(q,i,k) / (J11 * J22 - J21 * J12);
+         for (int j = 0; j < NC; j++)
+         {
+            const complex_t x1 = R.div_qpts(i,0,j,k);
```
Comment (YohannDudouit) [inline fem/tbilininteg.hpp]: Does the divergence needs to be stored? Why is it not computed on the fly with a tensor?

Comment (YohannDudouit) [reply fem/tbilininteg.hpp]: Does this operator works for non-H(div) basis? The operator should be able to compute the divergence for a H1 basis too, no?

File: fem/tbilininteg.hpp (hunk 6 above)
Comment (YohannDudouit) [reply fem/tbilininteg.hpp]: Ok, i see that the continuity mapping appears here. Shouldn't we have a more general MassIntegrator then?

File: fem/tbilininteg.hpp (hunk 7 above)
Comment (wang84) [reply fem/tbilininteg.hpp]: No, it is design for H(div) space. The main reason is that the mapping for different spaces are not the same.

File: fem/tbilininteg.hpp (hunk 3 above)
Comment (wang84) [reply fem/tbilininteg.hpp]: The mapping for H1, H(curl), and H(div) are different. But we can redesign and simplify the mapping, like diffusion kernel in H1 are the same with the mass kernel in H(curl).

File: fem/tevaluator.hpp [hunk 8]
```diff
@@ -1417,6 +1417,3255 @@ class FieldEvaluator
...
+
+   template <int D1, int D2, bool Add,
+             typename qpt_layout_t, typename qpt_data_t,
+             typename M_layout_t,  typename M_data_t,
+             typename D11_layout_t,typename D11_data_t,
+             typename D12_layout_t,typename D12_data_t,
+             typename D21_layout_t,typename D21_data_t,
+             typename D22_layout_t,typename D22_data_t>
+   MFEM_ALWAYS_INLINE
+   void AssembleBlock(const qpt_layout_t &qpt_layout,const qpt_data_t   &qpt_data,
```
Comment (YohannDudouit) [inline fem/tevaluator.hpp]: Would it be possible to have a few comments for this functions, even if I get the idea, it's really difficult to know what's really happening without any comment.

File: fem/tevaluator.hpp [hunk 9]
```diff
@@ -1417,6 +1417,3255 @@ class FieldEvaluator
...
+   // D_layout is (TDOF x TDOF x NumComp).
+   template <typename qpt_layout_t, typename qpt_data_t,
+             typename D_layout_t, typename D_data_t>
+   MFEM_ALWAYS_INLINE
+   void AssembleCurlCurl(const qpt_layout_t &qpt_layout,
+                         const qpt_data_t   &qpt_data,
+                         const D_layout_t   &D_layout,
+                         D_data_t           &D_data) const
+   {
+      //block (0,0)
```
Comment (YohannDudouit) [inline fem/tevaluator.hpp]: Reviewing this part feels a lot like the game of the seven errors. Would it be possible to factorize code, or add comments to understand the general idea... The order in which the blocks are build, what the variables are for instance.

**Comment by barker29** [comment]
@YohannDudouit I fully agree with your comments, but want to point out in defense of @wang84 that basically all of your concerns apply equally to the existing templated H1 code, which is also undocumented, duplicates itself, and is very difficult to read.

**Comment by tzanio** [comment]
I feel that this is superseded by #1204. Any thoughts if we should keep it or close it?

@barker29, @v-dobrev, @YohannDudouit, @dylan-copeland?

**Comment by barker29** [comment]
We can close this.

In my opinion, we should think about deprecating all the `miniapps/performance` code, because it is not being maintained and probably will not be maintained going forward.

**Comment by tzanio** [comment]
Thanks Andrew. @v-dobrev, @YohannDudouit, @dylan-copeland -- what do you think?

**Comment by dylan-copeland** [comment]
I have not begun reviewing the 7534 lines of code and will defer to others familiar with this PR. It appears to have H(div) implemented, but it would be "easy" to do H(div) based on PR #1204 now that it has been approved. In other words, the difficulty is not implementation but agreeing on how to incorporate it into mfem, and that has been done already in PR #1204.

**Comment by v-dobrev** [comment]
My guess is that the template code is noticeably faster on CPUs than the implementations using the `MFEM_FORALL` macros. It is possible to update the appropriate integrators to dispatch to the template code (instantiated for some set of parameters) when running on CPU.

File: fem/tbilininteg.hpp [hunk 10]
```diff
@@ -554,6 +583,1130 @@ struct TDiffusionKernel<3,3,complex_t>
...
+
+   template <typename IR, typename coeff_t, int NE>
+   struct CoefficientEval
+   {
+      typedef typename IntRuleCoefficient<IR,coeff_t,NE>::Type Type;
+   };
+
+   /** @brief Method used for un-assembled (matrix free) action.
+
+       @todo do not believe this equation description below is correct
```
Comment (camierjs) [inline fem/tbilininteg.hpp:1198]: We should address these comments through a couple of files.

File: fem/tbilininteg.hpp [hunk 11]
```diff
@@ -554,6 +583,1130 @@ struct TDiffusionKernel<3,3,complex_t>
...
+   /** @brief Method defining partial assembly.
+
+       The pointwise Dim x Dim matrices are
+       stored as symmetric (when asm_type == p_asm_data, i.e. A.layout.rank == 2)
+       or non-symmetric (when asm_type == f_asm_data, i.e. A.layout.rank == 3)
+       matrices.
+
+       A = (w/det(J))Jt*J
+
+       @todo believe the description for either this or 3D version is wrong
```
Comment (camierjs) [inline fem/tbilininteg.hpp:1248]: Here too.

File: fem/tbilininteg.hpp [hunk 12]
```diff
@@ -554,6 +583,1130 @@ struct TDiffusionKernel<3,3,complex_t>
...
+   /** @brief Method defining partial assembly.
+
+       The pointwise Dim x Dim matrices are
+       stored as symmetric (when asm_type == p_asm_data, i.e. A.layout.rank == 2)
+       or non-symmetric (when asm_type == f_asm_data, i.e. A.layout.rank == 3)
+       matrices.
+
+       A = (w/det(J)) JJ^t
+
+       @todo believe the description for either this or 2D version is wrong
```
Comment (camierjs) [inline fem/tbilininteg.hpp:1395]: Comment to update

**Pull request closed without merging**